docs/cn/curvefs_volume_space_design.md

```diff
@@ -0,0 +1,535 @@
+   
+   空间分配过程未记录具体分配给哪个 inode，在用 fsck 修复时，如何快速找回空间？(根据当前 client 写入了哪些 inode，分配了哪些 block group？)
+
+### 释放流程
+
+在将文件 truncate 变小、删除文件时触发空间的回收操作。
+
+#### s3 文件系统的实现
+
+在 truncate 后，空间的回收操作未进行处理，是利用 compaction 来处理，目前还有些场景无法处理，比如 100MiB 的文件 truncate 后，只新写入了 50MiB，那么后续的旧数据不会被清理。整体处理逻辑还是根据覆盖写 compaction 的逻辑来的。(https://github.com/opencurve/curve/issues/1049)
+
+删除文件的操作会在 metaserver 进行处理，其中删除对象的操作会在 3 副本上都执行。*这个逻辑需要修改，否则空间会被回收三次，如果其中一次回收后，又被分配出去，后两次的回收会导致其他文件的数据丢失。*
+
+#### 删除文件处理流程
+
+在 MetaServer 端删除文件、清理文件的使用空间。
+
+![curvefs space block group free](../images/curvefs-space-block-group-free.png)
+
+整体流程如上所示。
+
+在清理文件占用的空间时，可以先计算出占用空间属于哪些 block group，对于同一个 block group 的空间，进行合并回收，避免多次回收的开销。
+
+在确认需要释放 inode 占用的空间时，需要先获取到对应的 block group，然后再对 bitmap 进行清零操作，表示空间可用。
+
+#### 特殊场景
+
+1. 上述的流程中，假设需要回收空间所在的 block group 被其他 client 持有。其余可能存在的情况是
+   
+   a. block group 未被持有，所以不需要与 client 交互，回收 block group 的权限
+   
+   b. block group 被其他的 metaserver 持有。这种情况下，需要等待其他 metaserver 释放权限，而不是主动回收。
+
+2. 删除文件包括两个操作：清理空间和删除文件对应的 inode，如果在清理空间之后，删除 inode 之前，metaserver 进程挂掉。
+
+   在重启或 leader 切换后，会继续执行删除文件的操作，因此，会再次清理空间、删除 inode。所以，这里的问题是，第一次回收之后的空间，可能已经分配给了其他 inode，再次清理时，会导致其他文件的数据丢失。
+
+#### truncate 处理流程
+
+流程应该跟文件删除类似，只是操作的发起和执行是在 client 端，client 端需要获取对应 block group 的权限，然后执行空间释放操作。
+
+## 方案对比
+
+### 性能对比
+
+方案一中，Space Server 端负责细粒度的空间分配及持久化，所以当 client 端每次需要申请空间时，都需要发送 RPC 请求。
+
+方案二中，空间分配分为两个层级：粗粒度的空间分配由 MDS 进行，后续细粒度的空间分配在 client 内部处理，无需 RPC 交互。
+
+除 RPC 外，直观来看，持久化、幂等问题处理带来的开销应该是一致的。
+
+### 服务端异常情况下的影响
+
+方案一中，Space Server 在 IO 的核心路径上，所以服务异常会对 IO 产生影响，需要高可用来降低服务异常带来的影响。
+
+方案二中，空间分配的核心流程在 client 端内部处理，只有粗粒度的空间分配需要与 MDS 交互。整体来看，MDS 异常的情况下对 IO 影响较小。
+
+# 结论
+
+空间分配基于方案二进行。
+
+相较于方案一，方案二有以下几点优势:
+
+1. 不需要为空间分配服务重新设计高可用，只需要 MDS 提供高可用的服务即可（目前已实现）。
+2. 空间分配的核心流程在 client 端内部完成，在服务异常的情况下，影响较小。
+3. 空间分配信息直接持久化到卷上。
+
+方案中的遗留问题如下：
+
+1. 空间分配/回收过程中，更新 bitmap 和更新 inode 不是原子操作，再此过程中进程异常退出，会导致文件系统不一致。
+2. fuse-client 申请 block group 时可能会出现已完全分配的情况。
+
+## 第一阶段开发计划
+
+1. 完成基本的空间分配、回收功能
+2. 遗留问题先进行规避
+   1. 空间分配过程中进程异常退出后，先更新 bitmap，后更新 inode 会造成空间浪费的问题。**第一阶段先不处理**。
+   2. 空间回收过程中进程异常退出后，通过先更新 inode，再清理 bitmap 的方式，规避反向操作可能带来的不一致问题，造成的空间浪费问题，**第一阶段先不处理**。按照这种方式处理的时序图如下：
+![curvefs space block group free](../images/curvefs-space-block-group-free.png)
+
+   3. fuse-client 申请 block group 时，如果 block group 完全分配，进行主动的 extend 卷。
+   
+   
+### 相关 RPC 接口
+
+```cpp
+syntax="proto2";
+package curvefs.mds.space;
+option cc_generic_services=true;
+
+enum SpaceStatusCode {
+    SpaceOk = 0;
+    SpaceUnknownError = 1;
+    SpaceNoSpace = 2;
+    SpaceNotFound = 3;
+}
+
+enum BitmapLocation {
+    AtStart = 1;
+    AtEnd = 2;
+}
+
+message BlockGroup {
+    // start offset on volume
+    required uint64 offset = 1;
+    // total bytes of each block group
+    // this must be fixed when create fs or mkfs, so, this field maybe not needed
+    required uint32 size = 2;
+    // used bytes
+    required uint32 used = 3;
+    // available bytes
+    required uint32 available = 4;
```

Comment:
required uint32 available 
optional string owner

if available=false, there must be an owner?